app/models/derived_variables/shared_logic.rb

```diff
@@ -0,0 +1,27 @@
+module DerivedVariables::SharedLogic
+private
+
+  def reset_invalidated_derived_values!(dependencies)
+    dependencies.each do |dependency|
+      any_conditions_changed = dependency[:conditions].any? { |attribute, _value| send("#{attribute}_changed?") }
+      next unless any_conditions_changed
+
+      previously_in_derived_state = dependency[:conditions].all? { |attribute, value| send("#{attribute}_was") == value }
+      next unless previously_in_derived_state
+
+      dependency[:derived_values].each do |derived_attribute, _derived_value|
+        Rails.logger.debug("Cleared derived #{derived_attribute} value")
+        send("#{derived_attribute}=", nil)
+      end
+    end
+  end
+
+  def set_encoded_derived_values!(dependencies)
```

Comment:
whas does the word encoded mean in this context?

Reply:
I agree this is not named well.
This is where the situations where deriving happens are able to be encoded in the dependencies hash, would love to have an extended meta-code conversation about naming on this after lunch